lib/Fmt_ast.ml

```diff
@@ -811,21 +775,97 @@ and type_constr_and_body c xbody =
       , sub_exp ~ctx:exp_ctx exp )
   | _ -> (None, xbody)
 
-and fmt_modalities c modalities =
+and fmt_modalities ?(break = true) c modalities =
   let fmt_modality {txt= Modality modality; loc} =
     Cmts.fmt c loc (str modality)
   in
   if List.is_empty modalities then noop
-  else fmt "@ @@@@ " $ list modalities " " fmt_modality
+  else
+    fmt (if break then "@ " else " ")
+    $ fmt "@@@@ "
+    $ hvbox 0 (list modalities "@ " fmt_modality)
 
 and fmt_modes ~ats c modes =
   let fmt_mode {txt= Mode mode; loc} = Cmts.fmt c loc (str mode) in
   if List.is_empty modes then noop
   else
     let fmt_ats =
-      match ats with `One -> fmt "@ @@ " | `Two -> fmt "@ @@@@ "
+      match ats with
+      | `Zero -> fmt "@ "
+      | `One -> fmt "@ @@ "
+      | `Two -> fmt "@ @@@@ "
     in
-    fmt_ats $ list modes " " fmt_mode
+    fmt_ats $ hvbox 0 (list modes "@ " fmt_mode)
+
+and fmt_type_var ~have_tick c (s : ty_var) =
+  let {txt= name_opt; loc= name_loc}, jkind_opt = s in
+  ( Cmts.fmt c name_loc
+  @@
+  match name_opt with
+  | None -> str "_"
+  | Some var_name ->
+      fmt_if_k have_tick
+        ( str "'"
+        (* [' a'] is a valid type variable, the space is required to not lex
+           as a char. https://github.com/ocaml/ocaml/pull/2034 *)
+        $ fmt_if
+            (String.length var_name > 1 && Char.equal var_name.[1] '\'')
+            " " )
+      $ str var_name )
+  $ Option.value_map jkind_opt ~default:noop
+      ~f:(fmt_jkind_constr ~ctx:(Tyv s) c)
+
+and fmt_type_var_with_parenze ~have_tick c (s : ty_var) =
+  let jkind_annot = type_var_has_jkind_annot s in
+  cbox_if jkind_annot 0
+    (wrap_if jkind_annot "(" ")" (fmt_type_var ~have_tick c s))
+
+and fmt_jkind c ~ctx {txt= jkd; loc} =
+  let inner_ctx = Jkd jkd in
+  let parens, fmt =
+    match jkd with
+    | Default -> (false, fmt "_")
+    | Abbreviation abbrev -> (false, fmt_str_loc c abbrev)
+    | Mod (jkind, modes) ->
+        let parens = match ctx with Jkd (Product _) -> true | _ -> false in
+        let mode_fmt = hvbox 0 (fmt_modes ~ats:`Zero c modes) in
+        let fmt =
+          fmt_jkind c ~ctx:inner_ctx jkind
+          $ fmt "@ mod" $ Cmts.fmt_within c loc $ mode_fmt
+        in
+        (parens, fmt)
+    | With (jkind, type_) ->
+        let parens = match ctx with Jkd (Product _) -> true | _ -> false in
+        let types_fmt =
+          fmt_core_type c ~box:true (sub_typ ~ctx:inner_ctx type_)
+        in
+        let fmt =
+          fmt_jkind c ~ctx:inner_ctx jkind
+          $ fmt "@ with " $ Cmts.fmt_within c loc $ types_fmt
+        in
+        (parens, fmt)
+    | Kind_of type_ ->
+        let type_fmt =
+          fmt_core_type c ~box:true (sub_typ ~ctx:inner_ctx type_)
+        in
+        (false, fmt "kind_of_@ " $ Cmts.fmt_within c loc $ type_fmt)
+    | Product kinds ->
+        let parens =
+          match ctx with
+          | Jkd (Product _ | Mod _ | With _) -> true
+          | _ -> false
```

Comment:
Let's get rid of this catch-all case so we are forced to consider any new kinds.